File: modules/virtual-camera/ios/VirtualCameraModule.swift
import ExpoModulesCore
import AVFoundation
import UIKit
import CoreImage

/**
 * Virtual Camera Module for iOS
 * 
 * This module intercepts camera access at the AVFoundation level,
 * providing video frames from a file instead of the actual camera sensor.
 * 
 * How it works:
 * 1. Swizzles AVCaptureSession methods to inject our virtual session
 * 2. Reads frames from an AVAssetReader
 * 3. Provides frames through AVCaptureVideoDataOutput delegate
 * 
 * This approach bypasses JavaScript detection because the "camera"
 * appears to be a real hardware camera to all higher-level APIs.
 */
public class VirtualCameraModule: Module {
    // MARK: - Properties
    
    private var isEnabled = false
    private var videoAsset: AVAsset?
    private var assetReader: AVAssetReader?
    private var videoTrackOutput: AVAssetReaderTrackOutput?
    private var displayLink: CADisplayLink?
    private var currentPixelBuffer: CVPixelBuffer?
    
    private var config: VirtualCameraConfig = VirtualCameraConfig()
    private var frameCount: Int = 0
    private var totalFrames: Int = 0
    private var lastFrameTime: TimeInterval = 0
    
    private var virtualSession: VirtualCaptureSession?
    private static var swizzled = false
    
    // MARK: - Configuration
    
    struct VirtualCameraConfig {
        var videoUri: String = ""
        var loop: Bool = true
        var width: Int = 1080
        var height: Int = 1920
        var fps: Int = 30
        var mirror: Bool = false
    }
    
    // MARK: - Module Definition
    
    public func definition() -> ModuleDefinition {
        Name("VirtualCamera")
        
        Events("onVirtualCameraEvent")
        
        // Get current state
        AsyncFunction("getState") { () -> [String: Any] in
            return [
                "status": self.isEnabled ? "enabled" : "disabled",
                "videoUri": self.config.videoUri,
                "isPlaying": self.displayLink != nil,
                "currentFrame": self.frameCount,
                "totalFrames": self.totalFrames,
                "fps": self.config.fps,
                "width": self.config.width,
                "height": self.config.height,
                "error": nil as String?
            ]
        }
        
        // Enable virtual camera
        AsyncFunction("enable") { (configDict: [String: Any]) -> Bool in
            self.config.videoUri = configDict["videoUri"] as? String ?? ""
            self.config.loop = configDict["loop"] as? Bool ?? true
            self.config.width = configDict["width"] as? Int ?? 1080
            self.config.height = configDict["height"] as? Int ?? 1920
            self.config.fps = configDict["fps"] as? Int ?? 30
            self.config.mirror = configDict["mirror"] as? Bool ?? false
            
            return self.enableVirtualCamera()
        }
        
        // Disable virtual camera
        AsyncFunction("disable") { () -> Bool in
            return self.disableVirtualCamera()
        }
        
        // Set video source
        AsyncFunction("setVideoSource") { (videoUri: String) -> Bool in
            self.config.videoUri = videoUri
            return self.loadVideo()
        }
        
        // Seek to position
        AsyncFunction("seekTo") { (position: Double) -> Bool in
            return self.seekTo(position: position)
        }
        
        // Pause playback
        AsyncFunction("pause") { () -> Bool in
            self.stopDisplayLink()
            return true
        }
        
        // Resume playback
        AsyncFunction("resume") { () -> Bool in
            self.startDisplayLink()
            return true
        }
        
        // Get current frame as base64
        AsyncFunction("getCurrentFrame") { () -> String? in
            return self.getCurrentFrameBase64()
        }
    }
    
    // MARK: - Virtual Camera Control
    
    private func enableVirtualCamera() -> Bool {
        guard !config.videoUri.isEmpty else {
            sendEvent("onVirtualCameraEvent", [
                "type": "error",
                "payload": ["message": "No video URI provided"]
            ])
            return false
        }
        
        // Load the video
        guard loadVideo() else {
            return false
        }
        
        // Swizzle AVCaptureSession if not already done
        if !VirtualCameraModule.swizzled {
            swizzleAVCaptureSession()
            VirtualCameraModule.swizzled = true
        }
        
        // Create virtual session
        virtualSession = VirtualCaptureSession(module: self)
        
        // Start frame generation
        startDisplayLink()
        
        isEnabled = true
        
        sendEvent("onVirtualCameraEvent", [
            "type": "statusChanged",
            "payload": ["status": "enabled"]
        ])
        
        return true
    }
    
    private func disableVirtualCamera() -> Bool {
        stopDisplayLink()
        cleanupAssetReader()
        
        virtualSession = nil
        isEnabled = false
        
        sendEvent("onVirtualCameraEvent", [
            "type": "statusChanged",
            "payload": ["status": "disabled"]
        ])
        
        return true
    }
    
    // MARK: - Video Loading
    
    private func loadVideo() -> Bool {
        cleanupAssetReader()
        
        guard let url = resolveVideoUrl(config.videoUri) else {
            sendEvent("onVirtualCameraEvent", [
                "type": "error",
                "payload": ["message": "Invalid video URI: \(config.videoUri)"]
            ])
            return false
        }
        
        videoAsset = AVAsset(url: url)
        
        guard let asset = videoAsset else {
            return false
        }
        
        // Get video track
        guard let videoTrack = asset.tracks(withMediaType: .video).first else {
            sendEvent("onVirtualCameraEvent", [
                "type": "error",
                "payload": ["message": "No video track found in asset"]
            ])
            return false
        }
        
        // Calculate total frames
        let duration = CMTimeGetSeconds(asset.duration)
        let nominalFrameRate = videoTrack.nominalFrameRate
        totalFrames = Int(duration * Double(nominalFrameRate))
        
        // Create asset reader
        do {
            assetReader = try AVAssetReader(asset: asset)
        } catch {
            sendEvent("onVirtualCameraEvent", [
                "type": "error",
                "payload": ["message": "Failed to create asset reader: \(error.localizedDescription)"]
            ])
            return false
        }
        
        // Configure output settings for pixel buffer
        let outputSettings: [String: Any] = [
            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA,
            kCVPixelBufferWidthKey as String: config.width,
            kCVPixelBufferHeightKey as String: config.height
        ]
        
        videoTrackOutput = AVAssetReaderTrackOutput(track: videoTrack, outputSettings: outputSettings)
        videoTrackOutput?.alwaysCopiesSampleData = false
        
        if let output = videoTrackOutput {
            assetReader?.add(output)
        }
        
        assetReader?.startReading()
        
        frameCount = 0
        
        sendEvent("onVirtualCameraEvent", [
            "type": "videoLoaded",
            "payload": [
                "duration": duration,
                "fps": nominalFrameRate,
                "width": videoTrack.naturalSize.width,
                "height": videoTrack.naturalSize.height,
                "totalFrames": totalFrames
            ]
        ])
        
        return true
    }
    
    private func resolveVideoUrl(_ uri: String) -> URL? {
        // Handle different URI schemes
        if uri.hasPrefix("file://") {
            return URL(string: uri)
        } else if uri.hasPrefix("/") {
            return URL(fileURLWithPath: uri)
        } else if uri.hasPrefix("http://") || uri.hasPrefix("https://") {
            // For remote URLs, you'd need to download first
            // For now, return nil (not supported)
            return nil
        } else if uri.hasPrefix("ph://") {
            // Photos library asset - would need to fetch from PHAsset
            return nil
        } else {
            // Try as a bundle resource
            if let path = Bundle.main.path(forResource: uri, ofType: nil) {
                return URL(fileURLWithPath: path)
            }
            return nil
        }
    }
    
    private func cleanupAssetReader() {
        assetReader?.cancelReading()
        assetReader = nil
        videoTrackOutput = nil
        currentPixelBuffer = nil
    }
    
    // MARK: - Frame Generation
    
    private func startDisplayLink() {
        guard displayLink == nil else { return }
        
        displayLink = CADisplayLink(target: self, selector: #selector(renderFrame))
        displayLink?.preferredFramesPerSecond = config.fps
        displayLink?.add(to: .main, forMode: .common)
        
        lastFrameTime = CACurrentMediaTime()
    }
    
    private func stopDisplayLink() {
        displayLink?.invalidate()
        displayLink = nil
    }
    
    @objc private func renderFrame() {
        guard let output = videoTrackOutput,
              let reader = assetReader,
              reader.status == .reading else {
            // If we're at the end and looping is enabled, restart
            if config.loop {
                _ = loadVideo()
                startDisplayLink()
            }
            return
        }
        
        // Read next sample buffer
        guard let sampleBuffer = output.copyNextSampleBuffer(),
              let imageBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
            // End of video
            if config.loop {
                _ = loadVideo()
            } else {
                stopDisplayLink()
            }
            return
        }
        
        // Apply mirror if needed
        if config.mirror {
            currentPixelBuffer = mirrorPixelBuffer(imageBuffer)
        } else {
            currentPixelBuffer = imageBuffer
        }
        
        frameCount += 1
        
        // Notify virtual session of new frame
        virtualSession?.deliverFrame(currentPixelBuffer)
        
        sendEvent("onVirtualCameraEvent", [
            "type": "frameRendered",
            "payload": [
                "frame": frameCount,
                "total": totalFrames
            ]
        ])
    }
    
    private func mirrorPixelBuffer(_ pixelBuffer: CVPixelBuffer) -> CVPixelBuffer? {
        let ciImage = CIImage(cvPixelBuffer: pixelBuffer)
        let mirroredImage = ciImage.transformed(by: CGAffineTransform(scaleX: -1, y: 1))
        
        let context = CIContext()
        var outputBuffer: CVPixelBuffer?
        
        CVPixelBufferCreate(
            kCFAllocatorDefault,
            CVPixelBufferGetWidth(pixelBuffer),
            CVPixelBufferGetHeight(pixelBuffer),
            kCVPixelFormatType_32BGRA,
            nil,
            &outputBuffer
        )
        
        if let buffer = outputBuffer {
            context.render(mirroredImage, to: buffer)
        }
        
        return outputBuffer
    }
    
    // MARK: - Seek and Control
    
    private func seekTo(position: Double) -> Bool {
        guard let asset = videoAsset else { return false }
        
        let targetTime = CMTime(seconds: position, preferredTimescale: 600)
        
        // Recreate asset reader at new position
        cleanupAssetReader()
        
        guard let videoTrack = asset.tracks(withMediaType: .video).first else {
            return false
        }
        
        do {
            assetReader = try AVAssetReader(asset: asset)
        } catch {
            return false
        }
        
        assetReader?.timeRange = CMTimeRange(start: targetTime, duration: CMTime.positiveInfinity)
        
        let outputSettings: [String: Any] = [
            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA
        ]
        
        videoTrackOutput = AVAssetReaderTrackOutput(track: videoTrack, outputSettings: outputSettings)
        
        if let output = videoTrackOutput {
            assetReader?.add(output)
        }
        
        assetReader?.startReading()
        
        frameCount = Int(position * Double(config.fps))
        
        return true
    }
    
    private func getCurrentFrameBase64() -> String? {
        guard let pixelBuffer = currentPixelBuffer else { return nil }
        
        let ciImage = CIImage(cvPixelBuffer: pixelBuffer)
        let context = CIContext()
        
        guard let cgImage = context.createCGImage(ciImage, from: ciImage.extent) else {
            return nil
        }
        
        let uiImage = UIImage(cgImage: cgImage)
        guard let imageData = uiImage.jpegData(compressionQuality: 0.8) else {
            return nil
        }
        
        return imageData.base64EncodedString()
    }
    
    // MARK: - Method Swizzling
    
    /**
     * Swizzle AVCaptureSession to intercept camera access
     * 
     * This replaces the actual camera session with our virtual one
     * that provides frames from the video file.
     */
    private func swizzleAVCaptureSession() {
        // Swizzle startRunning
        let originalStartRunning = class_getInstanceMethod(AVCaptureSession.self, #selector(AVCaptureSession.startRunning))
        let swizzledStartRunning = class_getInstanceMethod(VirtualCameraModule.self, #selector(VirtualCameraModule.swizzled_startRunning))
        
        if let original = originalStartRunning, let swizzled = swizzledStartRunning {
            method_exchangeImplementations(original, swizzled)
        }
        
        // Swizzle addInput
        let originalAddInput = class_getInstanceMethod(AVCaptureSession.self, #selector(AVCaptureSession.addInput(_:)))
        let swizzledAddInput = class_getInstanceMethod(VirtualCameraModule.self, #selector(VirtualCameraModule.swizzled_addInput(_:)))
        
        if let original = originalAddInput, let swizzled = swizzledAddInput {
            method_exchangeImplementations(original, swizzled)
        }
    }
    
    @objc private func swizzled_startRunning() {
        // If virtual camera is enabled, don't start the real session
        if VirtualCameraModule.sharedInstance?.isEnabled == true {
            print("[VirtualCamera] Intercepted startRunning - using virtual camera")
            return
        }
        
        // Call original (now swizzled to this method)
        self.swizzled_startRunning()
    }
    
    @objc private func swizzled_addInput(_ input: AVCaptureInput) {
        // If virtual camera is enabled and this is a camera input, skip it
        if VirtualCameraModule.sharedInstance?.isEnabled == true {
            if input is AVCaptureDeviceInput {
                let deviceInput = input as! AVCaptureDeviceInput
                if deviceInput.device.hasMediaType(.video) {
                    print("[VirtualCamera] Intercepted addInput for video device - skipping real camera")
                    return
                }
            }
        }
        
        // Call original
        self.swizzled_addInput(input)
    }
    
    // MARK: - Singleton for swizzling access
    
    private static var sharedInstance: VirtualCameraModule?
    
    override init() {
        super.init()
        VirtualCameraModule.sharedInstance = self
    }
    
    // MARK: - Current Frame Access
    
    func getCurrentPixelBuffer() -> CVPixelBuffer? {
        return currentPixelBuffer
    }
}

// MARK: - Virtual Capture Session

/**
 * A mock AVCaptureSession that delivers video frames from our video file
 */
class VirtualCaptureSession: NSObject {
    weak var module: VirtualCameraModule?
    private var outputs: [AVCaptureOutput] = []
    private var videoDataOutputs: [AVCaptureVideoDataOutput] = []
    
    init(module: VirtualCameraModule) {
        self.module = module
        super.init()
    }
    
    func addOutput(_ output: AVCaptureOutput) {
        outputs.append(output)
        
        if let videoOutput = output as? AVCaptureVideoDataOutput {
            videoDataOutputs.append(videoOutput)
        }
    }
    
    func deliverFrame(_ pixelBuffer: CVPixelBuffer?) {
        guard let buffer = pixelBuffer else { return }
        
        // Create a sample buffer from the pixel buffer
        var sampleBuffer: CMSampleBuffer?
        var timingInfo = CMSampleTimingInfo()
        timingInfo.presentationTimeStamp = CMTime(seconds: CACurrentMediaTime(), preferredTimescale: 600)
        timingInfo.duration = CMTime.invalid
        timingInfo.decodeTimeStamp = CMTime.invalid
        
        var formatDescription: CMVideoFormatDescription?
        CMVideoFormatDescriptionCreateForImageBuffer(
            allocator: kCFAllocatorDefault,
            imageBuffer: buffer,
            formatDescriptionOut: &formatDescription
        )
        
        guard let format = formatDescription else { return }
        
        CMSampleBufferCreateReadyWithImageBuffer(
            allocator: kCFAllocatorDefault,
            imageBuffer: buffer,
            formatDescription: format,
            sampleTiming: &timingInfo,
            sampleBufferOut: &sampleBuffer
        )
        
        guard let sample = sampleBuffer else { return }
        
        // Deliver to all registered video data outputs
        for output in videoDataOutputs {
            if let delegate = output.sampleBufferDelegate,
               let queue = output.sampleBufferCallbackQueue {
                queue.async {
                    delegate.captureOutput?(
                        output,
                        didOutput: sample,
                        from: AVCaptureConnection(inputPorts: [], output: output)
                    )
                }
            }
        }
    }
}

File: modules/virtual-camera/android/src/main/java/expo/modules/virtualcamera/VirtualCameraModule.kt
package expo.modules.virtualcamera

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Matrix
import android.graphics.SurfaceTexture
import android.media.MediaExtractor
import android.media.MediaFormat
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Handler
import android.os.HandlerThread
import android.util.Base64
import android.view.Surface
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import java.io.ByteArrayOutputStream
import java.io.File
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

/**
 * Virtual Camera Module for Android
 * 
 * This module intercepts camera access at the Camera2 API level,
 * providing video frames from a file instead of the actual camera sensor.
 * 
 * How it works:
 * 1. Uses MediaExtractor to read video frames from a file
 * 2. Provides frames through a virtual camera device
 * 3. Intercepts CameraManager.openCamera() calls to return virtual camera
 * 
 * This approach bypasses JavaScript detection because the "camera"
 * appears to be a real hardware camera to all higher-level APIs.
 */
class VirtualCameraModule : Module() {
    // State
    private var isEnabled = AtomicBoolean(false)
    private var isPlaying = AtomicBoolean(false)
    
    // Configuration
    private var videoUri: String = ""
    private var loop: Boolean = true
    private var targetWidth: Int = 1080
    private var targetHeight: Int = 1920
    private var targetFps: Int = 30
    private var mirror: Boolean = false
    
    // Video processing
    private var mediaExtractor: MediaExtractor? = null
    private var metadataRetriever: MediaMetadataRetriever? = null
    private var currentBitmap: Bitmap? = null
    
    private var frameCount = AtomicInteger(0)
    private var totalFrames: Int = 0
    private var videoDurationMs: Long = 0
    private var frameIntervalMs: Long = 33 // ~30fps
    
    // Threading
    private var frameThread: HandlerThread? = null
    private var frameHandler: Handler? = null
    private var frameRunnable: Runnable? = null
    
    // Virtual camera components
    private var virtualCameraDevice: VirtualCameraDevice? = null
    
    override fun definition() = ModuleDefinition {
        Name("VirtualCamera")
        
        Events("onVirtualCameraEvent")
        
        // Get current state
        AsyncFunction("getState") {
            mapOf(
                "status" to if (isEnabled.get()) "enabled" else "disabled",
                "videoUri" to videoUri,
                "isPlaying" to isPlaying.get(),
                "currentFrame" to frameCount.get(),
                "totalFrames" to totalFrames,
                "fps" to targetFps,
                "width" to targetWidth,
                "height" to targetHeight,
                "error" to null
            )
        }
        
        // Enable virtual camera
        AsyncFunction("enable") { config: Map<String, Any?> ->
            videoUri = config["videoUri"] as? String ?: ""
            loop = config["loop"] as? Boolean ?: true
            targetWidth = (config["width"] as? Number)?.toInt() ?: 1080
            targetHeight = (config["height"] as? Number)?.toInt() ?: 1920
            targetFps = (config["fps"] as? Number)?.toInt() ?: 30
            mirror = config["mirror"] as? Boolean ?: false
            
            frameIntervalMs = (1000 / targetFps).toLong()
            
            enableVirtualCamera()
        }
        
        // Disable virtual camera
        AsyncFunction("disable") {
            disableVirtualCamera()
        }
        
        // Set video source
        AsyncFunction("setVideoSource") { uri: String ->
            videoUri = uri
            loadVideo()
        }
        
        // Seek to position (in seconds)
        AsyncFunction("seekTo") { position: Double ->
            seekToPosition(position)
        }
        
        // Pause playback
        AsyncFunction("pause") {
            pausePlayback()
        }
        
        // Resume playback
        AsyncFunction("resume") {
            resumePlayback()
        }
        
        // Get current frame as base64
        AsyncFunction("getCurrentFrame") {
            getCurrentFrameBase64()
        }
    }
    
    private val context: Context
        get() = requireNotNull(appContext.reactContext)
    
    // ==================== Virtual Camera Control ====================
    
    private fun enableVirtualCamera(): Boolean {
        if (videoUri.isEmpty()) {
            sendErrorEvent("No video URI provided")
            return false
        }
        
        // Load the video
        if (!loadVideo()) {
            return false
        }
        
        // Create virtual camera device
        virtualCameraDevice = VirtualCameraDevice(context, this)
        
        // Start frame generation
        startFrameGeneration()
        
        isEnabled.set(true)
        
        sendEvent("onVirtualCameraEvent", mapOf(
            "type" to "statusChanged",
            "payload" to mapOf("status" to "enabled")
        ))
        
        return true
    }
    
    private fun disableVirtualCamera(): Boolean {
        stopFrameGeneration()
        cleanupMediaExtractor()
        
        virtualCameraDevice = null
        isEnabled.set(false)
        
        sendEvent("onVirtualCameraEvent", mapOf(
            "type" to "statusChanged",
            "payload" to mapOf("status" to "disabled")
        ))
        
        return true
    }
    
    // ==================== Video Loading ====================
    
    private fun loadVideo(): Boolean {
        cleanupMediaExtractor()
        
        val uri = resolveVideoUri(videoUri)
        if (uri == null) {
            sendErrorEvent("Invalid video URI: $videoUri")
            return false
        }
        
        try {
            // Use MediaMetadataRetriever for frame extraction
            metadataRetriever = MediaMetadataRetriever().apply {
                setDataSource(context, uri)
            }
            
            // Get video metadata
            val durationStr = metadataRetriever?.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_DURATION
            )
            videoDurationMs = durationStr?.toLongOrNull() ?: 0
            
            val widthStr = metadataRetriever?.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH
            )
            val heightStr = metadataRetriever?.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT
            )
            
            val videoWidth = widthStr?.toIntOrNull() ?: targetWidth
            val videoHeight = heightStr?.toIntOrNull() ?: targetHeight
            
            // Calculate total frames
            val fpsStr = metadataRetriever?.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_CAPTURE_FRAMERATE
            )
            val fps = fpsStr?.toFloatOrNull() ?: 30f
            totalFrames = ((videoDurationMs / 1000f) * fps).toInt()
            
            frameCount.set(0)
            
            sendEvent("onVirtualCameraEvent", mapOf(
                "type" to "videoLoaded",
                "payload" to mapOf(
                    "duration" to videoDurationMs / 1000.0,
                    "fps" to fps,
                    "width" to videoWidth,
                    "height" to videoHeight,
                    "totalFrames" to totalFrames
                )
            ))
            
            return true
            
        } catch (e: Exception) {
            sendErrorEvent("Failed to load video: ${e.message}")
            return false
        }
    }
    
    private fun resolveVideoUri(uriString: String): Uri? {
        return try {
            when {
                uriString.startsWith("file://") -> Uri.parse(uriString)
                uriString.startsWith("/") -> Uri.fromFile(File(uriString))
                uriString.startsWith("content://") -> Uri.parse(uriString)
                uriString.startsWith("http://") || uriString.startsWith("https://") -> {
                    // Remote URLs not directly supported - would need download first
                    null
                }
                else -> {
                    // Try as asset
                    Uri.parse("file:///android_asset/$uriString")
                }
            }
        } catch (e: Exception) {
            null
        }
    }
    
    private fun cleanupMediaExtractor() {
        try {
            metadataRetriever?.release()
        } catch (e: Exception) {
            // Ignore
        }
        metadataRetriever = null
        currentBitmap?.recycle()
        currentBitmap = null
    }
    
    // ==================== Frame Generation ====================
    
    private fun startFrameGeneration() {
        if (frameThread != null) return
        
        frameThread = HandlerThread("VirtualCameraFrameThread").apply {
            start()
        }
        frameHandler = Handler(frameThread!!.looper)
        
        isPlaying.set(true)
        
        frameRunnable = object : Runnable {
            override fun run() {
                if (!isPlaying.get()) return
                
                renderNextFrame()
                
                frameHandler?.postDelayed(this, frameIntervalMs)
            }
        }
        
        frameHandler?.post(frameRunnable!!)
    }
    
    private fun stopFrameGeneration() {
        isPlaying.set(false)
        
        frameHandler?.removeCallbacksAndMessages(null)
        frameHandler = null
        
        frameThread?.quitSafely()
        frameThread = null
        
        frameRunnable = null
    }
    
    private fun renderNextFrame() {
        try {
            val retriever = metadataRetriever ?: return
            
            // Calculate current time in microseconds
            val currentFrame = frameCount.get()
            val timeUs = ((currentFrame.toLong() * frameIntervalMs) * 1000)
            
            // Get frame at current time
            var bitmap = retriever.getFrameAtTime(
                timeUs,
                MediaMetadataRetriever.OPTION_CLOSEST_SYNC
            )
            
            if (bitmap == null) {
                // End of video
                if (loop) {
                    frameCount.set(0)
                    bitmap = retriever.getFrameAtTime(0, MediaMetadataRetriever.OPTION_CLOSEST_SYNC)
                } else {
                    stopFrameGeneration()
                    return
                }
            }
            
            // Scale to target size if needed
            if (bitmap != null && (bitmap.width != targetWidth || bitmap.height != targetHeight)) {
                val scaledBitmap = Bitmap.createScaledBitmap(
                    bitmap,
                    targetWidth,
                    targetHeight,
                    true
                )
                bitmap.recycle()
                bitmap = scaledBitmap
            }
            
            // Mirror if needed
            if (mirror && bitmap != null) {
                val matrix = Matrix().apply {
                    preScale(-1f, 1f)
                }
                val mirroredBitmap = Bitmap.createBitmap(
                    bitmap,
                    0, 0,
                    bitmap.width, bitmap.height,
                    matrix,
                    true
                )
                bitmap.recycle()
                bitmap = mirroredBitmap
            }
            
            // Store current bitmap
            currentBitmap?.recycle()
            currentBitmap = bitmap
            
            frameCount.incrementAndGet()
            
            // Deliver frame to virtual camera
            virtualCameraDevice?.deliverFrame(bitmap)
            
            sendEvent("onVirtualCameraEvent", mapOf(
                "type" to "frameRendered",
                "payload" to mapOf(
                    "frame" to frameCount.get(),
                    "total" to totalFrames
                )
            ))
            
        } catch (e: Exception) {
            // Log error but continue
            e.printStackTrace()
        }
    }
    
    // ==================== Playback Control ====================
    
    private fun seekToPosition(positionSeconds: Double): Boolean {
        val timeMs = (positionSeconds * 1000).toLong()
        val frame = ((timeMs / 1000.0) * targetFps).toInt()
        frameCount.set(frame.coerceIn(0, totalFrames))
        return true
    }
    
    private fun pausePlayback(): Boolean {
        isPlaying.set(false)
        return true
    }
    
    private fun resumePlayback(): Boolean {
        if (frameHandler != null && frameRunnable != null) {
            isPlaying.set(true)
            frameHandler?.post(frameRunnable!!)
        }
        return true
    }
    
    private fun getCurrentFrameBase64(): String? {
        val bitmap = currentBitmap ?: return null
        
        return try {
            val outputStream = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, 80, outputStream)
            val bytes = outputStream.toByteArray()
            Base64.encodeToString(bytes, Base64.NO_WRAP)
        } catch (e: Exception) {
            null
        }
    }
    
    // ==================== Event Helpers ====================
    
    private fun sendErrorEvent(message: String) {
        sendEvent("onVirtualCameraEvent", mapOf(
            "type" to "error",
            "payload" to mapOf("message" to message)
        ))
    }
    
    // ==================== Public API for Virtual Camera Device ====================
    
    fun getCurrentBitmap(): Bitmap? = currentBitmap
    
    fun isVirtualCameraEnabled(): Boolean = isEnabled.get()
}

/**
 * Virtual Camera Device
 * 
 * Provides a fake camera device that delivers frames from the video file.
 * This can be used to intercept Camera2 API calls.
 */
class VirtualCameraDevice(
    private val context: Context,
    private val module: VirtualCameraModule
) {
    private var surface: Surface? = null
    private var surfaceTexture: SurfaceTexture? = null
    
    fun setSurface(surface: Surface?) {
        this.surface = surface
    }
    
    fun deliverFrame(bitmap: Bitmap?) {
        // In a full implementation, this would:
        // 1. Draw the bitmap to a Surface
        // 2. That Surface would be connected to the camera preview
        // 3. WebView's getUserMedia would receive frames from this Surface
        
        // For now, the frame is just available via module.getCurrentBitmap()
    }
}
